// @ingameltd/node-przelewy24/dist/enums/Country.js
"use strict";
/**
 * MIT License
 *
 * Copyright (c) 2019 Kasun Vithanage
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Country = void 0;
/**
 * Country Code
 *
 * @enum {string}
 */
var Country;
(function (Country) {
    Country["Andorra"] = "AD";
    Country["Austria"] = "AT";
    Country["Belgium"] = "BE";
    Country["Cyprus"] = "CY";
    Country["CzechRepublic"] = "CZ";
    Country["Denmark"] = "DK";
    Country["Estonia"] = "EE";
    Country["Finland"] = "FI";
    Country["France"] = "FR";
    Country["Greece"] = "EL";
    Country["Spain"] = "ES";
    Country["Norway"] = "NO";
    Country["Poland"] = "PL";
    Country["Portugal"] = "PT";
    Country["SanMarino"] = "SM";
    Country["Slovakia"] = "SK";
    Country["Slovenia"] = "SI";
    Country["Switzerland"] = "CH";
    Country["Sweden"] = "SE";
    Country["Hungary"] = "HU";
    Country["GreatBritain"] = "GB";
    Country["Italy"] = "IT";
    Country["Netherland"] = "NL";
    Country["Ireland"] = "IE";
    Country["Island"] = "IS";
    Country["Lithuania"] = "LT";
    Country["Latvia"] = "LV";
    Country["Luxemburg"] = "LU";
    Country["Malta"] = "MT";
    Country["USA"] = "US";
    Country["Canada"] = "CA";
    Country["Japan"] = "JP";
    Country["Ukraine"] = "UA";
    Country["Belarus"] = "BY";
    Country["Russia"] = "RU";
})(Country = exports.Country || (exports.Country = {}));
